### ISSUE TITLE
Implement the vm service protocol for the dart debugger

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/webdev#135]
TODO: list all the required methods (maybe @vsmenon?)

**Comment by jakemac53** [comment]
- [x] getVM
- [x] getIsolate
- [ ] callServiceExtension
- [ ] getVersion
- [ ] getScripts
- [ ] getObject
- [ ] streamListen
- [ ] addBreakpoint
- [ ] removeBreakpoint
- [ ] callMethod
- [ ] getStack
- [ ] invoke
- [ ] resume
- [ ] setExceptionPauseMode
- [ ] pause

**Comment by DanTup** [comment]
Some things VS Code currently uses that I don't think are covered above:

- setLibraryDebuggable
- evaluate
- evaluteInFrame

**Comment by jakemac53** [comment]
Ya, `evaluate` is probably deserving of its own issue entirely as that is going to be quite involved. We have an extremely rudimentary version now that just evals the expression as JS within the correct library scope (which was enough to unblock the inspector).

True support like a debugger needs will be a lot more involved though.

**Comment by jakemac53** [comment]
Also - do you have any sense of how often you will be running evals?

**Comment by DanTup** [comment]
> do you have any sense of how often you will be running evals?

Yeah, it's used by the watch window, the debug console (where the use can type expressions), and I think also when the user hovers over symbols in the editor (it just occurs to me that VS Code myst have some heuristics to device what expression under the mouse to send for evaluation, since I'm pretty sure we don't help it out here).

There was talk about potentially providing batch support for this, since it can be slow for Dart use too (in VS Code we have an option to execute all getters when you're examining an object, and currently that's an evaluate for each one). It's possible that some of VS Code's use of `evaluate` could be moved over to `invoke` (for ex. we use evaluate to just call `toString()` in some places for sure) so if there will likely be a perf difference between them, I'll move over what I can.

**Comment by DanTup** [comment]
Oh also, we use `evaluateInFrame` for conditional breakpoints and log points too. Both essentially just set breakpoints, then when they're hit we evaluate the expression/print and then immediately resume. This means conditional breakpoints/logpoints slow things down quite considerably (even before accounting for the evaluation). I think users tend to understand that though, and I don't think people use them that much.

**Comment by jakemac53** [comment]
I think for the watch window case at least caching should be able to save us to some degree - we can cache the compiled JS by expression/library, presumably the watch window is sending the exact same call repeatedly?

**Comment by DanTup** [comment]
As you step it will, yeah - I was thinking more about the having 15 watch variables, that it sends 15 requests (or if any are expanded objects or arrays, potentially more) sequentially.

**Comment by jakemac53** [comment]
Ah  ya - maybe we can resolve that case with some sort of batching mechanism.

In general its going to be very difficult to provide a good experience I think though.

I do wonder if we want some sort of special compiler mode for this..? Knowing absolutely nothing about ddc it seems like you could maybe do a "dumb" compile that doesn't require info about dependencies and uses all dynamic calls. Maybe that is a terrible idea or totally impossible idk, cc @jmesserly @nshahan @Markzipan for input :).

**Comment by DanTup** [comment]
Yeah, we've talked about batching elsewhere (possibly with @bkonyi?), but I don't think anything has been started yet.

I don't know the details, but I think Flutter mobile has a compiler service that handles evaluate requests (I think the on-device VM can't evaluate them so it sends them back to Flutter tools over a VM service to compile?). I'm not sure it helps you at all, though I think @aam worked on it so he may be able to explain it better if it sounds relevant.

**Comment by bkonyi** [comment]
Yes, we have plans to support batching for certain operations like `evaluate` and `invoke`, but I'm not sure what those will look like yet. Should be something I can get to soon though.

**Comment by alan-knight** [comment]
For a lot of uses it's possible to fake a batch evaluate by doing a single evaluate that returns a batch of things. But real support would be nice.

We have some code for doing evaluation internally, but it needs to be adapted for a different build system, module system, etc., and will probably come after things like breakpoint support.

**Comment by jmesserly** [comment]
> I do wonder if we want some sort of special compiler mode for this..? Knowing absolutely nothing about ddc it seems like you could maybe do a "dumb" compile that doesn't require info about dependencies and uses all dynamic calls.

yeah Alan and I have brainstormed this.

There's a few different approaches we could take:

**Approach 1**: compile the expression, defining all names as dynamic variables. For example:
```dart
// console/watch expression the user typed:
foo.bar.baz + qux
```
becomes:
```dart
// Note, we simply pass in all identifier-like names. If they weren't actually in the JS scope,
// we'll pass in an object with nSM that throws an appropriate error.
//
// We don't even need to parse the expression at all--false positives are fine.
evaluate(foo, bar, baz, qux) => foo.bar.baz + qux;
```

One problem with that is class names and library prefixes:
```dart
// Foo and Bar.baz are constructors, qux is a library prefix in scope:
Foo() + Bar.baz() + qux.f()
```
We could solve that by passing in wrapper objects, that understand how to call constructors (via nSM, or internal magic in DDC's dynamic dispatch) or lookup members in libraries. But that *still* doesn't work in all cases:
```dart
// Oops, now "Bar" needs to be a type available at compile time.
Foo<Bar>()
```

Another problem is mutation:
```dart
// attempt to set the "foo" variable/field/setter
foo = bar.baz
```

For that to work, we'll need to use getters/setters instead of parameters:
```dart
import 'special_repl_magic.dart' show scope;
// read/write against the current scope.
// may call getters/setters depending on what these names refer to
get foo => scope['foo'];
set foo(x) => scope['foo'] = x;
get bar => scope['bar'];
set bar(x) => scope['bar'] = x;
get baz => scope['baz'];
set baz(x) => scope['baz'] = x;

evaluate() => foo = bar.baz;
```

---

**Approach 2**: parse and evaluate the expression directly, rather than using the compiler.

Since we're using dynamic dispatch, we're already relying on most operations being interpreted at run time. If we do a real Dart parse of the expression, we could reliably perform lookups against the currently active scope. As a result, type arguments would work as expected. This would also make it much easier to declare and mutate local variables in the console (typically, this is accomplished by giving the console its own scope in the scope chain).

The benefit here is we can implement any features that users need; there's no upper limit. Responsiveness should be better too, in most cases.

With either approach, we'll be much worse if the user pastes a complex "for" loop that does local computation inside the REPL (without relying on already-compiled functions). The current approach, which uses the compiler and static types, would be superior for that use case. However, I believe that use case is very uncommon.

Another issue is type inference:

```dart
// list is a  List<int>
list.map((x) => x + 42).toList()
```

With approach 1, that will produce a `List<dynamic>`. Approach 2 will give the same result, unless we implement some type inference in the interpreter (note: this isn't quite as bad as it sounds; DDC's runtime already has some type inference code ported from CFE. It's used to provide better error messages.)

Personally I think `List<dynamic>` is an acceptable result. An easy workaround is `list.map((int x) => x + 42).toList()`

---

As to which approach is better: I think option 2 is better in the long run, but option 1 is very appealing in the short term. The first approach is quite close to what we're doing now, but it lets watch expressions work regardless of context they're evaluated in.

So my recommendation is to implement Option 1, and consider option 2 down the road if needed. 

Interested to hear everyone's thoughts on this. :)

**Comment by jmesserly** [comment]
I should mention for completeness sake, we also brainstormed a third approach:

**Approach 3**: keep a warm compile state, that is capable of quickly compiling a JS expression in any context (taking into account appropriate types).

Essentially, the compiler keeps around enough state to quickly re-enter a context and "pretend compile" an expression in that context.

I'm not sure how low we can get the latency, but we might be able to make it fast enough for watch expressions. We'd need to investigate DDC's backend and the common front end to assess whether this is really a feasible approach. It might turn out to be more work than Approach 2.

**Comment by Markzipan** [comment]
Thanks for the breakdown, Jenny! I'm a fan of strategy #2 if I'm
understanding it correctly. Are you saying we run a simple Dart interpreter
to cut out the slower codegen part?

#3 sounds fun in an experimental way (but would need a spike).

**Comment by jmesserly** [comment]
yeah, strategy 2 is nice because it'd be fast to evaluate expressions, and we don't have any limits as to what we can support.

**Comment by alan-knight** [comment]
See also #352

**Comment by grouma** [comment]
The vm service protocol is mostly implemented for the use cases that we require. This issue doesn't provide much value. We have duplicate issues for specific subsets of the vm protocol, e.g. https://github.com/dart-lang/webdev/issues/591. Going to close this one.

**Comment by DanTup** [comment]
`setLibraryDebuggable` doesn't seem to be implemented (at least when running with the unforked version), but VS Code uses this. Are there plans to support it? (for VS Code, we set all SDK libraries and non-local packages as not-debuggable by default).

```
[12:14:43 PM] [Observatory] [Info] UnimplementedError
#0      ChromeProxyService.setLibraryDebuggable (package:dwds/src/services/chrome_proxy_service.dart:386:5)
#1      VmServerConnection._delegateRequest (package:vm_service/vm_service.dart:1075:51)
<asynchronous suspension>
#2      _rootRunUnary (dart:async/zone.dart:1132:38)
#3      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
```

**Comment by grouma** [comment]
We already do some default black boxing so this isn't a huge concern. Should be straightforward to implement though. See https://github.com/dart-lang/webdev/issues/606
